diff a/src/hotspot/os_cpu/linux_x86/os_linux_x86.hpp b/src/hotspot/os_cpu/linux_x86/os_linux_x86.hpp
--- a/src/hotspot/os_cpu/linux_x86/os_linux_x86.hpp
+++ b/src/hotspot/os_cpu/linux_x86/os_linux_x86.hpp
@@ -30,14 +30,10 @@
 
   static jlong rdtsc();
 
   static bool is_allocatable(size_t bytes);
 
-  // Used to register dynamic code cache area with the OS
-  // Note: Currently only used in 64 bit Windows implementations
-  static bool register_code_area(char *low, char *high) { return true; }
-
   /*
    * Work-around for broken NX emulation using CS limit, Red Hat patch "Exec-Shield"
    * (IA32 only).
    *
    * Map and execute at a high VA to prevent CS lazy updates race with SMP MM
